crates/volta-core/src/session.rs

```diff
@@ -116,36 +114,30 @@ impl Session {
     /// - If it exists but doesn't have a Yarn version, then we merge the two,
     ///   pulling Yarn from the user default platform, if available
     /// - If there is no Project platform, then we use the user Default Platform
-    pub fn current_platform(&self) -> Fallible<Option<SourcedPlatformSpec>> {
-        if let Some(platform) = self.project_platform()? {
-            if platform.yarn.is_some() {
-                Ok(Some(SourcedPlatformSpec::project(platform)))
-            } else {
-                let default_yarn = self.default_platform()?.and_then(|p| p.yarn.clone());
-                let merged = Rc::new(PlatformSpec {
-                    node_runtime: platform.node_runtime.clone(),
-                    npm: platform.npm.clone(),
-                    yarn: default_yarn,
-                });
-                Ok(Some(SourcedPlatformSpec::merged(merged)))
+    pub fn current_platform(&self) -> Fallible<Option<Rc<dyn PlatformSpec>>> {
+        match self.project_platform()? {
+            Some(platform) => {
+                if platform.yarn.is_none() {
+                    if let Some(default) = self.default_platform()? {
+                        return Ok(Some(platform.merge(default)));
+                    }
+                }
+                Ok(Some(platform))
             }
-        } else if let Some(platform) = self.default_platform()? {
-            Ok(Some(SourcedPlatformSpec::default(platform)))
-        } else {
-            Ok(None)
+            None => match self.default_platform()? {
+                Some(platform) => Ok(Some(platform)),
+                None => Ok(None),
+            },
```

Comment:
Huh. There's an interesting quirk here where the match triggers implicit behavior I would not expect—one of the bits of fallout from the `match` ergonomics improvements I 98% really appreciate. I would normally expect that this would be identical to just doing `None => self.default_platform()`, but no.

Reply:
Yeah, I think I tried that as well. I believe it's a result of the fact that the Trait object is nested within other types. So it's not necessarily valid that `Fallible<Option<T>>` is assignable to `Fallible<Option<U>>` if `T: U`, but if you unwrap and rewrap it, it works out, since you are then explicit about how it's contained.